server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java

```diff
@@ -0,0 +1,132 @@
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.index.query.QueryShardException;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+/** Mapper for the doc_count field. */
+public class DocCountFieldMapper extends MetadataFieldMapper {
+
+    public static final String NAME = "_doc_count";
+    public static final String CONTENT_TYPE = "_doc_count";
+
+    public static final TypeParser PARSER = new ConfigurableTypeParser(
+        c -> new DocCountFieldMapper(),
+        c -> new DocCountFieldMapper.Builder());
+
+    static class Builder extends MetadataFieldMapper.Builder {
+
+        Builder() {
+            super(NAME);
+        }
+
+        @Override
+        protected List<Parameter<?>> getParameters() {
+            return Collections.emptyList();
+        }
+
+        @Override
+        public DocCountFieldMapper build(BuilderContext context) {
+            return new DocCountFieldMapper();
+        }
+    }
+
+    public static final class DocCountFieldType extends MappedFieldType {
+
+        public static final DocCountFieldType INSTANCE = new DocCountFieldType();
+
+        public DocCountFieldType() {
+            super(NAME, false, false, true, TextSearchInfo.NONE,  Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            return new DocValuesFieldExistsQuery(NAME);
+        }
+
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            throw new QueryShardException(context, "Field [" + name() + " ]of type [" + CONTENT_TYPE + "] is not searchable");
+        }
+    }
+
+    private DocCountFieldMapper() {
+        super(DocCountFieldType.INSTANCE);
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        if (context.parser().currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            Long value = context.parser().longValue(false);
+
+            if (value != null) {
+                if (value.longValue() <= 0) {
+                    throw new IllegalArgumentException("Field [" + fieldType().name() + "] must be a positive integer.");
+                }
+                final Field docCount = new NumericDocValuesField(NAME, value.longValue());
+                context.doc().add(docCount);
+            }
+        } else {
+            throw new IllegalArgumentException("Field [" + fieldType().name() + "] must be a positive integer.");
+        }
+    }
+
+    @Override
+    public void preParse(ParseContext context) { }
+
+    @Override
+    public ValueFetcher valueFetcher(MapperService mapperService, SearchLookup lookup, String format) {
```

Comment:
We **just** moved this method to `MappedFieldType`, so this will have to be moved to `DocCountFieldType` before merging. Also it would be great to add a test -- the most common place to do that is in a field type test like `BooleanFieldTypeTests#testFetchSourceValue`.